### PR TITLE
[RLLib] MARWIL and BC Config

### PR DESCRIPTION
**Pull request by ghost**
## Checks

- [ Y] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ Y] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [Y ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/trainer_config.py [hunk 1]
```diff
@@ -927,7 +927,7 @@ def offline_data(
         Returns:
             This updated TrainerConfig object.
         """
-        if input is not None:
+        if input_ is not None:
```
Comment (sven1977) [inline rllib/agents/trainer_config.py:930]: nice.

File: rllib/algorithms/marwil/bc.py [hunk 2]
```diff
@@ -1,25 +1,58 @@
...
+
+class BCConfig(MARWILConfig):
+    """Defines a BCConfig configuration class based on MARWILConfig
+    from which a new configuration can be built
+
+    Example:
+        >>> from ray.rllib.agents.marwil import BCConfig
+        >>> # Run this from the ray directory root.
+        >>> config = MARWILConfig().training(beta=1.0, lr=0.00001, gamma=0.99)\
+        ...             .offline_data(input_="./rllib/tests/data/cartpole/large.json")
```
Comment (sven1977) [inline rllib/algorithms/marwil/bc.py:15]:
```suggestion
        >>> config = BCConfig().training(lr=0.00001, gamma=0.99)\
        ...             .offline_data(input_="./rllib/tests/data/cartpole/large.json")
```

File: rllib/algorithms/marwil/bc.py [hunk 3]
```diff
@@ -1,25 +1,58 @@
...
+        >>> config.training(lr=tune.grid_search([0.001, 0.0001]), beta=0.75)
+        >>> # Set the config object's data path.
+        >>> # Run this from the ray directory root.
+        >>> config.offline_data(input_="./rllib/tests/data/cartpole/large.json")
+        >>> # Set the config object's env, used for evaluation.
+        >>> config.environment(env="CartPole-v0")
+        >>> # Use to_dict() to get the old-style python config dict
+        >>> # when running with tune.
+        >>> tune.run(
+        ...     "MARWIL",
```
Comment (sven1977) [inline rllib/algorithms/marwil/bc.py:37]:
```suggestion
        ...     "BC",
```

File: rllib/algorithms/marwil/marwil.py [hunk 4]
```diff
@@ -23,79 +26,160 @@
...
-    "bc_logstd_coeff": 0.0,
-
-    # === Parallelism ===
-    "num_workers": 0,
-})
-# __sphinx_doc_end__
-# fmt: on
+
+class MARWILConfig(TrainerConfig):
+    """Defines a MARWILTrainer configuration class from which a MARWILTrainer can be built.
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:31]:
```suggestion
    """Defines a configuration class from which a MARWILTrainer can be built.
```

File: rllib/algorithms/marwil/marwil.py [hunk 5]
```diff
@@ -23,79 +26,160 @@
...
+        moving_average_sqd_adv_norm_update_rate: Optional[float] = None,
+        moving_average_sqd_adv_norm_start: Optional[float] = None,
+        replay_buffer_size: Optional[int] = None,
+        learning_starts: Optional[int] = None,
+        vf_coeff: Optional[float] = None,
+        grad_clip: Optional[float] = None,
+        use_gae: Optional[bool] = None,
+        **kwargs,
+    ) -> "MARWILConfig":
+        """Sets the training related configuration.
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:119]: Can we add an empty line between headline and `Args:`? Same before `Returns:`.

File: rllib/algorithms/marwil/marwil.py [hunk 6]
```diff
@@ -23,79 +26,160 @@
...
+            (single and independent) timesteps.
+            The buffer gets filled by reading from the input files line-by-line and
+            adding all timesteps on one line at once. We then sample uniformly
+            from the buffer (`train_batch_size` samples) for each training step.
+            learning_starts: Number of steps to read before learning starts.
+            bc_logstd_coeff: A coefficient to encourage higher action distribution
+            entropy for exploration.
+            grad_clip: If specified, clip the global norm of gradients by this amount.
+            use_gae: If true, use the Generalized Advantage Estimator (GAE)
+            with a value function, see https://arxiv.org/pdf/1506.02438.pdf
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:139]: indent all by one tab:
e.g.
```
    grad_clip: blablabla ....
        more blablabla
```

File: rllib/offline/estimators/__init__.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+from ray.rllib.offline.estimators.importance_sampling import ImportanceSampling
```
Comment (sven1977) [inline rllib/offline/estimators/__init__.py:1]: Nice catch!

**Comment by gjoliver** [comment]
can you rebase? there still seems to be some legit test failures.

File: rllib/agents/maddpg/maddpg.py [hunk 8]
```diff
@@ -124,7 +124,7 @@
 # fmt: on
 
 
-def before_learn_on_batch(multi_agent_batch, policies, train_batch_size):
+def before_learn_on_batch(multi_agent_batch, policies, train_batch_size, framework):
```
Comment (sven1977) [inline rllib/agents/maddpg/maddpg.py:127]: Could you remove these changes here? They are probably from another PR?

**Comment by rapotdar** [comment]
Sorry about that, should be fixed now

**Review by sven1977 — APPROVED** [review]
Looks good. Thanks for this nice PR @rapotdar !
Some tests are failing. Let me take a quick look. ...

File: rllib/algorithms/marwil/marwil.py [hunk 9]
```diff
@@ -146,10 +192,10 @@ def training(
             )
         if moving_average_sqd_adv_norm_start is not None:
             self.moving_average_sqd_adv_norm_start = moving_average_sqd_adv_norm_start
-        if replay_buffer_size is not None:
-            self.replay_buffer_size = replay_buffer_size
-        if learning_starts is not None:
-            self.learning_starts = learning_starts
+        if replay_buffer_config is not None:
+            self.replay_buffer_config = replay_buffer_config
+        if use_gae is not None:
```
Comment (rapotdar) [inline rllib/algorithms/marwil/marwil.py:197]: I think use_gae needs to be removed from the config, per our discussion?

**Pull request merged**